File: SnakeLink_v2.0_Complete/software/hlink_protocol.py
"""
H-Link Protocol v2.0 — Python Implementation
"""

import struct
import time
from enum import IntEnum
from typing import Optional, Tuple
from dataclasses import dataclass


class PacketType(IntEnum):
    TEXT = 0x01
    FILE_CHUNK = 0x02
    ACK = 0x03
    HEARTBEAT = 0x04
    GPS = 0x05
    BATTERY = 0x06
    VOICE = 0x07
    VOICE_REQ = 0x08
    VOICE_ACK = 0x09
    EMERGENCY = 0x0A
    LINK_STATUS = 0x0B
    FILE_META = 0x0C
    ENCRYPTION = 0x0D
    CMD = 0x0E
    NACK = 0x0F
    PING = 0x10
    PONG = 0x11


class PacketFlags(IntEnum):
    NONE = 0x00
    ENCRYPTED = 0x01
    PRIORITY = 0x02
    FRAGMENT = 0x04
    LAST_FRAG = 0x08
    COMPRESSED = 0x10
    BROADCAST = 0x20


def crc8_itu(data: bytes) -> int:
    """CRC-8 ITU polynomial 0x07."""
    crc = 0x00
    for byte in data:
        crc ^= byte
        for _ in range(8):
            crc = ((crc << 1) ^ 0x07) if (crc & 0x80) else (crc << 1)
    return crc & 0xFF


@dataclass
class HLinkPacket:
    """H-Link packet structure."""
    sync: bytes = b"\xAA\x55"
    type: int = 0
    seq: int = 0
    timestamp: int = 0
    flags: int = 0
    length: int = 0
    payload: bytes = b""
    crc: int = 0

    _seq_counter = 0

    @classmethod
    def next_sequence(cls) -> int:
        cls._seq_counter = (cls._seq_counter + 1) & 0xFFFF
        return cls._seq_counter

    def compute_crc(self) -> int:
        return crc8_itu(self.payload)

    def is_valid(self) -> bool:
        if self.sync != b"\xAA\x55":
            return False
        if self.length > 240:
            return False
        return self.crc == self.compute_crc()

    def serialize(self) -> bytes:
        """Serialize packet to bytes."""
        self.crc = self.compute_crc()
        header = struct.pack(">BBHIBH",
            self.sync[0], self.sync[1],
            self.type & 0xFF,
            self.seq & 0xFFFF,
            self.timestamp & 0xFFFFFFFF,
            self.flags & 0xFF,
            self.length & 0xFFFF
        )
        return header + self.payload + bytes([self.crc])

    def deserialize(self, data: bytes) -> bool:
        """Deserialize bytes to packet."""
        if len(data) < 12:  # Minimum header size
            return False

        if data[0:2] != b"\xAA\x55":
            return False

        self.sync = data[0:2]
        self.type = data[2]
        self.seq = struct.unpack(">H", data[3:5])[0]
        self.timestamp = struct.unpack(">I", data[5:9])[0]
        self.flags = data[9]
        self.length = struct.unpack(">H", data[10:12])[0]

        if self.length > 240:
            return False

        if len(data) < 12 + self.length + 1:
            return False

        self.payload = data[12:12 + self.length]
        self.crc = data[12 + self.length]

        return self.is_valid()

    @classmethod
    def create(cls, ptype: PacketType, payload: bytes = b"", 
               flags: int = PacketFlags.NONE) -> "HLinkPacket":
        """Factory method to create a packet."""
        pkt = cls()
        pkt.type = ptype
        pkt.seq = cls.next_sequence()
        pkt.timestamp = int(time.time())
        pkt.flags = flags
        pkt.length = min(len(payload), 240)
        pkt.payload = payload[:pkt.length]
        pkt.crc = pkt.compute_crc()
        return pkt

    def type_name(self) -> str:
        try:
            return PacketType(self.type).name
        except:
            return f"UNKNOWN(0x{self.type:02X})"

File: SnakeLInk_v2.1_Complete/firmware/src/hlink_protocol.cpp
/**
 * @file hlink_protocol.cpp
 * @brief H-Link v2.0 Protocol Implementation
 */

#include "hlink_protocol.h"
#include <Arduino.h>

namespace hlink {

// ============================================================================
// CRC-8 TABLE
// ============================================================================
uint8_t CRC8::crcTable[256];
bool CRC8::tableInitialized = false;

void CRC8::generateTable() {
    if (tableInitialized) return;
    for (int i = 0; i < 256; i++) {
        uint8_t crc = i;
        for (int j = 0; j < 8; j++) {
            crc = (crc & 0x80) ? ((crc << 1) ^ CRC8_POLYNOMIAL) : (crc << 1);
        }
        crcTable[i] = crc;
    }
    tableInitialized = true;
}

uint8_t CRC8::compute(const uint8_t* data, size_t len) {
    if (!tableInitialized) generateTable();
    uint8_t crc = CRC8_INITIAL;
    for (size_t i = 0; i < len; i++) {
        crc = crcTable[crc ^ data[i]];
    }
    return crc;
}

uint8_t CRC8::compute(const Packet& pkt) {
    return compute(pkt.payload, pkt.header.length);
}

// ============================================================================
// PACKET METHODS
// ============================================================================
static uint16_t globalSequence = 0;

Packet::Packet() {
    clear();
}

Packet::Packet(PacketType type, uint16_t seq, const uint8_t* data, uint16_t len, uint8_t flags) {
    clear();
    header.sync[0] = HLINK_SYNC_BYTE_1;
    header.sync[1] = HLINK_SYNC_BYTE_2;
    header.type = static_cast<uint8_t>(type);
    header.seq = seq;
    header.timestamp = 0;  // Set by sender
    header.flags = flags;
    header.length = (len > HLINK_MAX_PAYLOAD) ? HLINK_MAX_PAYLOAD : len;
    if (data && header.length > 0) {
        memcpy(payload, data, header.length);
    }
    computeCRC();
}

void Packet::clear() {
    memset(&header, 0, sizeof(header));
    memset(payload, 0, sizeof(payload));
    crc = 0;
}

void Packet::setPayload(const uint8_t* data, uint16_t len) {
    header.length = (len > HLINK_MAX_PAYLOAD) ? HLINK_MAX_PAYLOAD : len;
    if (data && header.length > 0) {
        memcpy(payload, data, header.length);
    }
}

void Packet::computeCRC() {
    crc = CRC8::compute(payload, header.length);
}

bool Packet::verifyCRC() const {
    if (header.length > HLINK_MAX_PAYLOAD) return false;
    uint8_t computed = CRC8::compute(payload, header.length);
    return computed == crc;
}

bool Packet::isValid() const {
    return (header.sync[0] == HLINK_SYNC_BYTE_1) && 
           (header.sync[1] == HLINK_SYNC_BYTE_2) &&
           (header.length <= HLINK_MAX_PAYLOAD) &&
           verifyCRC();
}

bool Packet::isSync(const uint8_t* buffer) {
    return (buffer[0] == HLINK_SYNC_BYTE_1) && (buffer[1] == HLINK_SYNC_BYTE_2);
}

int Packet::serialize(uint8_t* buffer, size_t bufferSize) const {
    size_t total = totalSize();
    if (bufferSize < total) return -1;

    size_t idx = 0;
    buffer[idx++] = header.sync[0];
    buffer[idx++] = header.sync[1];
    buffer[idx++] = header.type;
    buffer[idx++] = (header.seq >> 8) & 0xFF;
    buffer[idx++] = header.seq & 0xFF;
    buffer[idx++] = (header.timestamp >> 24) & 0xFF;
    buffer[idx++] = (header.timestamp >> 16) & 0xFF;
    buffer[idx++] = (header.timestamp >> 8) & 0xFF;
    buffer[idx++] = header.timestamp & 0xFF;
    buffer[idx++] = header.flags;
    buffer[idx++] = (header.length >> 8) & 0xFF;
    buffer[idx++] = header.length & 0xFF;

    if (header.length > 0) {
        memcpy(&buffer[idx], payload, header.length);
        idx += header.length;
    }
    buffer[idx++] = crc;

    return static_cast<int>(idx);
}

int Packet::deserialize(const uint8_t* buffer, size_t bufferSize) {
    if (bufferSize < HLINK_HEADER_SIZE + 1) return 0;
    if (!isSync(buffer)) return 0;

    size_t idx = 0;
    header.sync[0] = buffer[idx++];
    header.sync[1] = buffer[idx++];
    header.type = buffer[idx++];
    header.seq = (static_cast<uint16_t>(buffer[idx]) << 8) | buffer[idx+1]; idx += 2;
    header.timestamp = (static_cast<uint32_t>(buffer[idx]) << 24) |
                       (static_cast<uint32_t>(buffer[idx+1]) << 16) |
                       (static_cast<uint32_t>(buffer[idx+2]) << 8) |
                       static_cast<uint32_t>(buffer[idx+3]); idx += 4;
    header.flags = buffer[idx++];
    header.length = (static_cast<uint16_t>(buffer[idx]) << 8) | buffer[idx+1]; idx += 2;

    if (header.length > HLINK_MAX_PAYLOAD) return 0;
    if (bufferSize < idx + header.length + 1) return 0;  // Need more data

    if (header.length > 0) {
        memcpy(payload, &buffer[idx], header.length);
        idx += header.length;
    }
    crc = buffer[idx++];

    return isValid() ? static_cast<int>(idx) : 0;
}

const char* Packet::typeToString(PacketType type) {
    switch (type) {
        case TYPE_TEXT: return "TEXT";
        case TYPE_FILE_CHUNK: return "FILE_CHUNK";
        case TYPE_ACK: return "ACK";
        case TYPE_HEARTBEAT: return "HEARTBEAT";
        case TYPE_GPS: return "GPS";
        case TYPE_BATTERY: return "BATTERY";
        case TYPE_VOICE: return "VOICE";
        case TYPE_VOICE_REQ: return "VOICE_REQ";
        case TYPE_VOICE_ACK: return "VOICE_ACK";
        case TYPE_EMERGENCY: return "EMERGENCY";
        case TYPE_LINK_STATUS: return "LINK_STATUS";
        case TYPE_FILE_META: return "FILE_META";
        case TYPE_ENCRYPTION: return "ENCRYPTION";
        case TYPE_CMD: return "CMD";
        case TYPE_NACK: return "NACK";
        case TYPE_PING: return "PING";
        case TYPE_PONG: return "PONG";
        default: return "UNKNOWN";
    }
}

uint16_t Packet::nextSequence() {
    globalSequence++;
    if (globalSequence > HLINK_SEQ_MAX) globalSequence = 1;
    return globalSequence;
}

// ============================================================================
// ARQ MANAGER
// ============================================================================
ARQManager::ARQManager(uint16_t to, uint8_t retries) 
    : timeoutMs(to), maxRetries(retries), pending(false), pendingSeq(0), 
      lastSendTime(0), currentSlot(0) {
    for (auto& q : queue) q.reset();
}

bool ARQManager::sendPacket(const Packet& pkt) {
    if (pending) return false;  // Stop-and-wait: only one pending at a time

    // Find free slot
    size_t slot = currentSlot;
    queue[slot].packet = pkt;
    queue[slot].sendTime = millis();
    queue[slot].retryCount = 0;
    queue[slot].pending = true;

    pending = true;
    pendingSeq = pkt.header.seq;
    lastSendTime = millis();
    currentSlot = (currentSlot + 1) % QUEUE_SIZE;

    return true;
}

bool ARQManager::onAckReceived(uint16_t seq) {
    if (!pending || seq != pendingSeq) return false;

    for (auto& q : queue) {
        if (q.pending && q.packet.header.seq == seq) {
            q.reset();
            pending = false;
            pendingSeq = 0;
            return true;
        }
    }
    return false;
}

bool ARQManager::onNackReceived(uint16_t seq) {
    for (auto& q : queue) {
        if (q.pending && q.packet.header.seq == seq) {
            q.retryCount = maxRetries;  // Force immediate retransmit on next update
            return true;
        }
    }
    return false;
}

void ARQManager::update() {
    if (!pending) return;

    uint32_t now = millis();
    for (auto& q : queue) {
        if (!q.pending) continue;

        if (now - q.sendTime >= timeoutMs) {
            q.retryCount++;
            if (q.retryCount > maxRetries) {
                // Packet failed after max retries
                q.reset();
                pending = false;
                pendingSeq = 0;
                // TODO: Notify link manager of failure
            } else {
                // Retransmit
                q.sendTime = now;
                lastSendTime = now;
                // The actual retransmit is handled by caller checking hasPending()
            }
        }
    }
}

void ARQManager::reset() {
    for (auto& q : queue) q.reset();
    pending = false;
    pendingSeq = 0;
}

} // namespace hlink
